core/src/main/java/com/linecorp/armeria/common/MediaTypeSet.java

```diff
@@ -0,0 +1,345 @@
+/*
+ *  Copyright 2017 LINE Corporation
+ *
+ *  LINE Corporation licenses this file to you under the Apache License,
+ *  version 2.0 (the "License"); you may not use this file except in compliance
+ *  with the License. You may obtain a copy of the License at:
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ *  License for the specific language governing permissions and limitations
+ *  under the License.
+ */
+package com.linecorp.armeria.common;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.util.AbstractSet;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.Set;
+
+import com.google.common.base.Ascii;
+import com.google.common.base.Splitter;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Iterators;
+
+/**
+ * An immutable {@link Set} of {@link MediaType}s which provides useful methods for content negotiation.
+ *
+ * <p>This {@link Set} provides {@link #match(MediaType...)} and {@link #matchHeaders(CharSequence...)}
+ * so that a user can find the preferred {@link MediaType} that matches the specified media ranges. For example:
+ * <pre>{@code
+ * MediaTypeSet set = new MediaTypeSet(MediaType.HTML_UTF_8, MediaType.PLAIN_TEXT_UTF_8);
+ *
+ * Optional<MediaType> negotiated1 = set.matchHeaders("text/html; q=0.5, text/plain");
+ * assert negotiated1.isPresent();
+ * assert negotiated1.get().equals(MediaType.PLAIN_TEXT_UTF_8);
+ *
+ * Optional<MediaType> negotiated2 = set.matchHeaders("audio/*, text/*");
+ * assert negotiated2.isPresent();
+ * assert negotiated2.get().equals(MediaType.HTML_UTF_8);
+ *
+ * Optional<MediaType> negotiated3 = set.matchHeaders("video/webm");
+ * assert !negotiated3.isPresent();
+ * }</pre>
+ */
+public final class MediaTypeSet extends AbstractSet<MediaType> {
+
+    private static final String WILDCARD = MediaType.ANY_TYPE.type();
+    private static final String Q = "q";
+    private static final String Q_UPPERCASED = Ascii.toUpperCase(Q);
```

Comment:
Just realized `MediaType` lower-cases type, subtype and parameter names. No need for case-insensitive comparison for parameter names.